Allow time condition windows to cross midnight.

jaharkes opened the pull request:
**Description:**
Allow time conditions to cross the midnight boundary.

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
condition:
  - condition: time
     after: "23:00:00"
     before: "01:00:00"
```

**Checklist:**

If the code does not interact with devices:
- [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] Tests have been added to verify that the new code works.

jaharkes committed:
Allow time condition windows to cross midnight.

balloob commented on homeassistant/helpers/condition.py:
```diff
@@ -224,15 +224,34 @@ def template_if(hass, variables=None):
-        return False
 
-    if after is not None and now_time < after:
+    if not _in_time_window(now.time(), before, after):
```
Any reason why we would extract this into a new function and not inline it?

jaharkes replied:
I started off inline and got myself confused with the logic, I was propagating the in_window boolean way too far down. Splitting off the time window test from the date test helped get everything working correctly. I can probably fold it back in now.

jaharkes replied:
Still gets pretty messy when trying to fold it into a single function because the before/after tests can return either true or false depending on in_window.

Actually I'm starting to think I still have the logic wrong, if `now => before` evaluates to true we immediately return and we skip the corresponding after test even when before and after got swapped.

balloob commented on tests/helpers/test_condition.py:
```diff
@@ -66,3 +67,12 @@ def test_or_condition(self):
         self.hass.states.set('sensor.temperature', 100)
         assert test(self.hass)
+
+    def test_time_window(self):
```
The test should probably patch `dt_util.now()` like [this](https://github.com/home-assistant/home-assistant/blob/dev/tests/components/automation/test_time.py#L259-L260) so we know what we are testing.

jaharkes replied:
it shouldn't matter because we simply test that now is either between 6am and 6pm or between 6pm and 6am.

I guess it could fail when we test as the second ticks over from 06:00:00 to 06:00:01 between the two tests, but I'm honestly not that worried about that because the window is much smaller than a second.

jaharkes committed:
Address comments.

Fold _in_time_window back into the time() condition test.
Use specific time values to test the time window.

balloob commented:
🎉 💃 🐬
